controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -433,21 +448,105 @@
                     });
     }
 
-    @Override
-    public CompletableFuture<Map<String, Data>> getTxnInEpoch(int epoch) {
+    private CompletableFuture<Map<String, Data>> getTxnInEpoch(int epoch) {
         return Futures.exceptionallyExpecting(store.getChildren(getEpochPath(epoch)),
                 e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, Collections.emptyList())
                       .thenCompose(txIds -> Futures.allOfWithResults(txIds.stream().collect(
-                              Collectors.toMap(txId -> txId, txId -> Futures.exceptionallyExpecting(store.getData(getActiveTxPath(epoch, txId)),
+                              Collectors.toMap(txId -> txId, txId -> Futures.exceptionallyExpecting(store.getData(getActiveTxPath(UUID.fromString(txId))),
                                       e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, EMPTY_DATA)))
                               ).thenApply(txnMap -> txnMap.entrySet().stream().filter(x -> !x.getValue().equals(EMPTY_DATA))
                                                           .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)))
                       );
     }
 
     @Override
-    CompletableFuture<Version> createNewTransaction(final int epoch, final UUID txId, final byte[] txnRecord) {
-        final String activePath = getActiveTxPath(epoch, txId.toString());
+    public CompletableFuture<List<Map.Entry<UUID, ActiveTxnRecord>>> getOrderedCommittingTxnInLowestEpoch() {
```

Comment:
This one looks like an expensive call. Is it going to be a problem?

Reply:
As long as number of outstanding transcations to be committed for a stream are not too high, it will be ok.

Reply:
Ok, so this is O(n) where n is the number of outstanding transactions for a stream.

Reply:
yes, however we can limit it to a fixed upper bound without breaking any guarantees. 
like say we terminate whenever we have collected 1000 transactions. So the order in average case will approximately be closer to the upper bound.